hostapd: always use sae_password for mesh/SAE-auth

yogo1212 opened the pull request:
This patch enables the use of passwords that are exactly 64 characters
long with SAE.
With SAE, there is no distinction based on length like there is for PSK
(the PMK is generated per handshake, so everything is a password).

Fixes: #11324

yogo1212 committed:
hostapd: always use sae_password for mesh/SAE auth

This patch fixes a corner case when using passwords that are exactly 64
characters in length with mesh mode or passwords longer than 63 characters
with SAE because 'psk' is used instead of 'sae_password'.
SAE is obligatory for 802.11s (mesh point).

The 'psk' option for hostapd is suited for WPA2 and enforces length
restrictions on passwords. Values of 64 characters are treated as PMKs.
With SAE, PMKs are always generated during the handshake and there are no
length restrictions.
The 'sae_password' option is more suited for SAE and should be used
instead.

Before this patch, the 'sae_password' option is only used with mesh mode
passwords that are not 64 characters long.
As a consequence:
- mesh passwords can't be 64 characters in length
- SAE only works with passwords with lengths >8 and <=63 (due to psk
  limitation).

Fix this by always using 'sae_password' with SAE/mesh and applying the PMK
differentiation only when PSK is used.

Fixes: #11324
Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>
[ improve commit description ]
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

yogo1212 committed:
hostapd: add quotes in assignments

It's generally advised to use quotes for variable assignments in bash.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

Ansuel commented:
@yogo1212 I'm a bit confused by the commit description

For commit 1 I would ask you to reword it and give some reason. Something like "fix shellcheck warning... it's safer to..."

For commit 2 from commit description say that SAE have exactly 64 bit pass but in the code you do the opposite...
Can you better explain it? Also this cause a behavior change... Can mesh have a password lower than 64 char?

yogo1212 commented:
@Ansuel I've added a 'before this patch' section to the second commit.

With the more general phrasing for commit #1 (quotes in shell assignment), I felt that other such assignments should be quoted as well and added them.

yogo1212 commented:
@Ansuel 
> For commit 2 from commit description say that SAE have exactly 64 bit pass but in the code you do the opposite...
> Can you better explain it? Also this cause a behavior change... Can mesh have a password lower than 64 char?

Oops! I completely misunderstood you there. Unless WPA2 compatibility is desired, there are no length restrictions on SAE passphrases. They are just arbitrary input to a hashing function.

Ansuel commented on package/network/services/hostapd/files/hostapd.sh:
```diff
@@ -1372,14 +1372,12 @@ wpa_supplicant_add_network() {
 			key_mgmt="$wpa_key_mgmt"
 
-			if [ ${#key} -eq 64 ]; then
+			if [ "$_w_mode" = "mesh" -o "$auth_type" = "sae" ]; then
```
Please use || instead of -o

Ansuel replied:
(in case you answer all the question and this became the final code... please use || instead of -o

Ansuel commented:
To reduce delta can we use something like this??

```
			if [ ${#key} -eq 64 ]; then
				passphrase="psk=${key}"
			else
				if [ "$_w_mode" = "mesh" ]; then
					passphrase="sae_password=\"${key}\""
				elif [ "$auth_type" = "sae" ]; then
					passphrase="sae_password=\"${key}\""
				else
					passphrase="psk=\"${key}\""
				fi
			fi
```

yogo1212 commented:
@Ansuel thanks for your feedback!
and sorry for the confusion :-(

> Or the problem is in using psk with sae auth type? Why does mesh also require the sae_password? Also the commit still have the new line between fixes and sob... Also it's still confusing... "This patch enables the use of passwords that are exactly 64 characters long with SAE."
> In this commit you do exactly the opposite by removing the lock of 64 char...
> And you don't say anything about the mesh change and why it's also required.

The differentiation of 64-character PMKs from regular passphrases only makes sense for regular PSK authentication (because the PMK is connection-specific for SAE).
The `psk=` option is limited insofar as it does the interpretation of those 64-character passphrases and limits the length of passphrases ([wpa_supplicant.conf](https://w1.fi/cgit/hostap/tree/wpa_supplicant/wpa_supplicant.conf#n1095)).
wpa_supplicant falls back to the `psk` option with SAE if `sae_password` is not specified which is why a switch wasn't necessary for Openwrt.
But really, `sae_password` should be used with SAE (and, thus, mesh mode because 802.11s requires SAE to be used for authentication).

Without this change, wpa_supplicant is fed the `psk=` option with SAE when the `key` option UCI is 64 characters long - ironically. This breaks authentication for SAE (and mesh mode) with 64 character long passphrases even if they are perfectly legal.
The fault is that the check for 64-character-PMKs takes precedence over mesh mode detection. Also, SAE isn't handled at all in the original code.

To fix that, I moved the mesh check out of the branch so that it takes precedence over the 64-character-check and added a check for SAE to it so that the PMK-differentiation only takes place if `auth_type` is `psk`.
Arguably, the PMK differentiation could happen inside a branch to have a more visual emphasis - thinking about it, I will do that.

The PMK differentiation itself needs to stay because there is no key derivation without the `"`.

I hope this clears things up :-)

Ansuel commented:
@yogo1212 Ok this is much better and thanks for the clarification!

It would be good to put these info in the commit description.
Also you should still fix the extra new line between "Fixes" and "Signed-off-by" (remove the line between the 2 tag)

Maybe some hint like:

```
Currently in the case where SAE is used and a 64char password, psk option is wrongly used. This cause problems since....
This is wrong because sae_pass should always used with SAE mode and Mesh mode ...

Fix this by reworking the logic and always use sae_pass for SAE and mesh.....
... and correctly auth in these special mode.
```

Hope it's clear how the commit description should be improved. The title is correct.

yogo1212 commented:
@Ansuel pew! i've updated the description.

> Hope it's clear how the commit description should be improved

tbh, it still isn't entirely clear to me but feel free to amend the description as you wish - it will likely take less of your time than further explanations :-)

Ansuel commented:
@yogo1212 Ok for you this new description?

yogo1212 commented:
@Ansuel yes, thank you!
that is easier to read than my version.

the first sentence contained a few relics from editing and one technical vagueness.
i'm pushing a proposal.

Ansuel commented:
@yogo1212 nice merging this!
